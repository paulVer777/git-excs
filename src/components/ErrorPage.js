import React from 'react'
import {Link} from 'react-router-dom'

const ErrorPage = () => (
    <div>Error Page <Link to='/'>Home</Link></div>
)

export default ErrorPage