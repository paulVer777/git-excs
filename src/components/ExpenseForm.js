import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'


export default class ExpenseForm extends React.Component {

constructor(props){
    super(props)
    this.state = {
     description: props.expense ? props.expense.description : '',
     note: props.expense ? props.expense.note : '',
     amount: props.expense ? (props.expense.amount / 100).toString() : '',
     createdAt:props.expense ? moment(props.expense.createdAt) : moment(),
     calendarFocused:false,
     error:undefined
    }
}

setDescription = (event) => { // this = parent scope

    const description = event.target.value
    this.setState((prevState) => ({description}))

}
setNote = (event) => {
    const note = event.target.value
    this.setState((prevState) => ({note}))
}
setAmount = (event) => {
    const amount = event.target.value

    // amount.match(/^\d{1,}(\.\d{0,2})?$/) && this.setState((prevState) => ({amount}))
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
        this.setState((prevState) => ({amount}))
    }
}
onDateChange = (createdAt) => { // that function takes as a argument moment object
    this.setState((prevState) => ({createdAt}))
}
onFocusChange = ({focused}) =>{
 this.setState((prevState) => ({ calendarFocused:focused }))
}
onSubmitHandler = (event) => {
    event.preventDefault()
    if(!this.state.description || !this.state.amount ){
    this.setState((prevState) => ({error:'Please provide description and amount'}))
    }
    else{
    this.setState((prevState) => ({error:undefined}))
    
    this.props.onSubmit({
        description:this.state.description,
        note:this.state.note,
        amount:parseFloat(this.state.amount),
        createdAt:this.state.createdAt.valueOf()
    })
    }
}

render(){
    return (
        <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit = {this.onSubmitHandler}>
        <input type ='text' placeholder ='Description' autoFocus value = {this.state.description}
        onChange ={this.setDescription}
        />
        <input type ='text' placeholder ='amount'
        value = {this.state.amount}
        onChange ={this.setAmount}
        />
        <SingleDatePicker 
        date = {this.state.createdAt}
        onDateChange = {this.onDateChange}
        focused = {this.state.calendarFocused}
        onFocusChange = {this.onFocusChange}
        numberOfMonths = {1}
        isOutsideRange = {() => false} // all days will be avaivable to choose, including past
        />
        <textarea placeholder ='Add a note to your expense' value ={this.state.note}
        onChange ={this.setNote}
        />
        <button>Add expense</button>
        </form>
        </div>
    )}
}

