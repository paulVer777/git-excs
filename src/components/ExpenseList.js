import React from 'react'
import {connect} from 'react-redux' //connects component to the redux store
import ListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = (props) => ( // {...value } it destructures the props object so we pass variables instead of whole object , it looks like - description:'asdad' amount:'fdsff'
// dispatch    
<div> 
     
    {props.expenses.length === 0 ? (

        <p>You dont have any expenses yet</p>
    )
    :
    (
        props.expenses.map((value,index) => <ListItem key={value.id} {...value} />
    )
    )
    }
    </div>
)

// connect returns a function , and we need to call this (thats why we have "() after () " ) function and pass our component in as a argument
// In connect function we need to pass information we want to connect, we dont need the whole store, we just need a subset. 
// so we pass function as a argument and this function lets us determine what information from the store we want to grab 
// in that function we pass state as a argument and return an object on this object we can put any key value pairs we like 
// usually they are going to be things from the state (everything we put on this object will be accesible to a component we want to connect with 
//( through props))



const mapStateToProps = (state) => ( //when the store changes this callback function automatically rerun providing fresh values to us
    {                                // when you connect a component to the redux store its reactive which means as the store changes
        expenses:getVisibleExpenses(state.expenses,state.filter)     // our component is going to re render new values to the screen
    }
)

export default connect(mapStateToProps)(ExpenseList)


