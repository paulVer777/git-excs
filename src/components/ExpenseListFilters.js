import React from 'react'
import { connect } from 'react-redux';
import {setFilterText} from '../actions/filters'
import {sortByDate, sortByAmount, setEndDate, setStartDate} from '../actions/filters'
import {DateRangePicker} from 'react-dates'

export class ExpenseListFilters extends React.Component {
 constructor(props){
    super(props)
    this.state = {
        calendarFocused:null
    }
 }
onDatesChange = ({startDate, endDate}) => { // this function is going to be called by the react dates library
                        // its going to be caled with an object and on that object we are going to have 
                        //an start date and end date. (we destructure this object explicitly)
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
}
onFocusChange = (calendarFocused) => {
this.setState((prevState) => ({calendarFocused}))
}

onTextChange = (e) => this.props.setFilterText(e.target.value)

onSortChange = (e) => e.target.value === 'date' ? this.props.sortByDate : this.props.sortByAmount


render(){

return (

    <div>
    <input type = 'text' value = {this.props.filters.text} onChange = {this.onTextChange} />
    <select onChange = {this.onSortChange} value = {this.props.filters.sortBy} >
    <option value = 'date' >By date</option>
    <option value = 'amount' >By amount</option>
    </select>
    <DateRangePicker
    startDate = {this.props.filters.startDate}
    endDate = {this.props.filters.endDate}
    onDatesChange = {this.onDatesChange} //what we do when the dates change
    focusedInput = {this.state.calendarFocused}
    onFocusChange = {this.onFocusChange}
    numberOfMonths = {1}
    isOutsideRange = {()=> false}
    showClearDates = {true}
    />
    </div>

)}
}
   
const mapStateToProps = (state) => ({
    filters:state.filter
})


const mapDispatchToProps = (dispatch) => ({

    setFilterText:(text) => dispatch(setFilterText(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
})

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters)

