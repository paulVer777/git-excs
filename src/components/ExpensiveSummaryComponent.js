import React from 'react'
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses'
import {sumUp} from '../selectors/totality'


export const ExpensiveSummaryComponent = (props) => {

    const correctWord = props.numberOfExpenses === 1 ? 'expense' : 'expenses'

    return (
     <div>
        <p>{`Viewing ${props.numberOfExpenses } ${correctWord} with total amount of ${props.sumOfExpenses}`}</p>
    </div>
)

}
   


const mapStateToProps = (state) => (
    {
        sumOfExpenses:sumUp(getVisibleExpenses(state.expenses,state.filter)),
        numberOfExpenses:getVisibleExpenses(state.expenses,state.filter).length
    }
)

export default connect(mapStateToProps)(ExpensiveSummaryComponent)