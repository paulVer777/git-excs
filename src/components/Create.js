import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {addExpense} from '../actions/expenses'

export class Create extends React.Component {

    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.history.push(`/`)
    }

    render(){
        return (
    <div>
        <h1>Create page</h1>
        <ExpenseForm
            onSubmit = {this.onSubmit}
        />
    </div>
        )
    }
}
    
const mapDispatchToProps = (dispatch) => ({ //we pulled out the functions from components and passed it here in order to avoid redefining these functions every time component renders

    addExpense:(expense) => dispatch(addExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(Create)