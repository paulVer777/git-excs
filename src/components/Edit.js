import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm';
import {editExpense} from '../actions/expenses'
import {removeExpense} from '../actions/expenses'

export class Edit extends React.Component { 

    onSubmit = (expenseObj) => {

        this.props.editExpense(expenseObj,this.props.expense.id)
        this.props.history.push('/')
    }
    
    onClickHandlerRemove = () => {
        this.props.removeExpense(this.props.match.params.id)
        this.props.history.push('/')
    }

    render(){

        return(
            <div>
                <ExpenseForm
    
            expense = {this.props.expense}
            onSubmit = {this.onSubmit}
    
                />
                <button onClick = {this.onClickHandlerRemove}> Remove </button>     
        </div>
            )
    }
}
   
const mapDispatchToProps = (dispatch) => ({
    editExpense:( expenseObj,id ) => dispatch(editExpense(id,expenseObj)),
    removeExpense: (id) => dispatch(removeExpense({id}))

})

const mapStateToProps = (state,props) => ({
    expense: state.expenses.find((value,index) => value.id === props.match.params.id)
})

export default connect(mapStateToProps, mapDispatchToProps)(Edit)