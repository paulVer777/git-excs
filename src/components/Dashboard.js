import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensiveSummaryComponent from './ExpensiveSummaryComponent';

const Dashboard = (props) => (
    <div>
    <ExpenseListFilters/>
    <ExpensiveSummaryComponent />
    <ExpenseList/>
    
    </div>
)

export default Dashboard