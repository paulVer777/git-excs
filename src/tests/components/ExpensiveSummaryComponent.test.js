import React from 'react'
import { shallow} from 'enzyme'
import { ExpensiveSummaryComponent } from '../../components/ExpensiveSummaryComponent'

test('Should render ExpensiveSummaryComponent correctly', () => {

    const wrapper = shallow( <ExpensiveSummaryComponent /> )
    expect(wrapper).toMatchSnapshot()
})


test('Should render ExpensiveSummaryComponent with provided data correctly', () => {

    const wrapper = shallow( <ExpensiveSummaryComponent sumOfExpenses = {3000}  numberOfExpenses = {1} /> )
    expect(wrapper).toMatchSnapshot()
})