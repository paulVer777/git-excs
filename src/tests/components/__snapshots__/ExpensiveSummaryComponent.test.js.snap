// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensiveSummaryComponent correctly 1`] = `
<div>
  <p>
    Viewing undefined expenses with total amount of undefined
  </p>
</div>
`;

exports[`Should render ExpensiveSummaryComponent with provided data correctly 1`] = `
<div>
  <p>
    Viewing 1 expense with total amount of 3000
  </p>
</div>
`;
